en/cpp/api_changes.md

```diff
@@ -21,9 +21,116 @@ This means that breaking changes to the API result in a bump of the major versio
+});
+
+discover_future.wait();
+```
+
+**New way of discovering a system:**
+```cpp
+Mavsdk mavsdk;
+
+auto new_system_promise = std::promise<std::shared_ptr<System>>{};
+auto new_system_future = new_system_promise.get_future();
+mavsdk.subscribe_on_new_system([&mavsdk, &new_system_promise]() {
+    new_system_promise.set_value(mavsdk.systems().at(0));
+    mavsdk.subscribe_on_new_system(nullptr);
+});
+
+auto system = new_system_future.get();
+```
+
+To be notified about a system timing out later, or being discovered again, you can now use:
+```cpp
+system->subscribe_is_connected([](bool is_connected) {
+    if (is_connected) {
+        std::cout << "System has been discovered" << std::endl;
+    } else {
+        std::cout << "System has timed out" << std::endl;
+    }
+});
+```
+
+If you prefer to use sync APIs / polling, you can use this to wait for a vehicle and connect:
+```cpp
+Mavsdk mavsdk;
+
+while (mavsdk.systems().size() == 0) {
+    std::this_thread::sleep_for(std::chrono::seconds(1));
+}
+
+auto system = mavsdk.systems().at(0);
+```
+
+
+### Accessing systems
+
+We aim to make the API of MAVSDK as easy and safe to use as possible.
+However, as we'll see below, this still comes with trade-offs.
+
+When designing the API we decided to expose the MAVSDK user a reference to a system `System&`, which means the system will always be instantiated and there won't be any segfaults trying to access a system that is null.
+This mostly worked fine, however, had some drawbacks:
+- Without an actual system discovered the `System&` is basically a null reference that won't really work.
+  This is not really obvious unless you check its system and component IDs for 0.
+  This is not very intuitive and also means additional checks inside the plugins are required.
+- The ownership model is not very clear as `Mavsdk` owns the `System` but then each plugin only seems to require the `System`.
+  Internally, the plugins still require `Mavsdk` to be alive and working but that's not apparent through the API.
+- If there are multiple sytems they need to be accessed using the `uuid`.
+  This seems like a crutch because essentially we would just like to have a `std::vector` of systems but of course that's not posible with references.
+
+As the API for the uid was about the change anyway (see above) it seemed time to redesign the API to access systems with these considerations in mind.
+
+The basic change is to move away from the reference `System&` to a shared pointer `std::shared_ptr<System>`.
+
+This brings the following advantages:
+- No more confusing null reference systems.
+- Clearer ownership model as a `System` is now shared between `Mavsdk` and the plugins.
+- Nice getter `std::vector<std::shared_ptr<System>> systems()` allowing easier access to all systems.
+
+**Previous way to get a system:**
+```cpp
+System& system = mavsdk.system();
+```
+
+**New way to get a system:**
+```cpp
+std::shared_ptr<System> system = mavsdk.systems().at(0);
```

Comment:
Should this be getting the system by uuid? Could use a list instead of a vector. 
```suggestion
std::shared_ptr<System> system = mavsdk.systems().at(uuid);
```

Reply:
I don't have a strong opinion on that, except that I never used it this way :see_no_evil:. `mavsdk.systems()` is the interface to the discovered systems. So unless you hardcoded the `uuid` somewhere in your script, at this point you don't know what it will be. Typically, to know it you'll have to go through all the systems (usually there aren't so many) and check them individually.

And there, I have personally never had a use-case where I check the uuid. I do check if it has an autopilot (`system->has_autopilot()`), which tells me that it is a drone. When it isn't, I assume it is a ground station.

How do you use the uuid?

Reply:
>So unless you hardcoded the uuid somewhere in your script, at this point you don't know what it will be

This is exactly what I am doing :joy: 

I don't have strong opinions either way, it just feels more intuitive to grab a system by ID rather than by index. If you have more mavlink components than a single FC and a single GCS, you are going to be manually choosing `system_id`s anyway. Choosing by index would mean you have to select it first and then check the ID, or you must know in which order they enumerate.

Reply:
> it just feels more intuitive to grab a system by ID rather than by index

Oh right. I would agree with that. But the point here is that you just get a list of systems, and if you expect more than one you should go through them and do your own logic (whether it uses the uuid or flags like `has_autopilot`). Here the `systems().at(0)` is a way to do `systems().giveMeTheOnlySystemThere()` :laughing:.

Reply:
yeah this makes sense, keeps the API minimal